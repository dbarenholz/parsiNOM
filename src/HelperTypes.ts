import { Parser, STypeBase } from './Parser';

export type TupleToUnion<T extends readonly unknown[]> = T[number];

export type DeParser<T extends Parser<unknown>> = T extends Parser<infer U> ? U : never;

export type DeParserArray<T extends readonly Parser<unknown>[]> = {
	[I in keyof T]: T[I] extends Parser<infer R> ? R : never;
};
