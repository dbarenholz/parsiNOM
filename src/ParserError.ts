import { ParseFailure } from './HelperTypes';

function generateMessage(str: string, parseFailure: ParseFailure, verbose: boolean): string {
	const expectedMessage = `Expected ${parseFailure.expected.join(' or ')}`;
	let message = `Parse Failure: ${expectedMessage} at index ${parseFailure.furthest.index}, line ${parseFailure.furthest.line}, column ${parseFailure.furthest.column}.`;

	if (verbose) {
		const lines = str.split('\n');
		const failedLine = lines[parseFailure.furthest.line - 1]; // line is a one based index

		const linePrefix = `${parseFailure.furthest.line} |   `;
		message += `\n\n${linePrefix}${failedLine}`;
		message += `\n${' '.repeat(parseFailure.furthest.column - 1 + linePrefix.length)}^ (${expectedMessage})`;
	}

	return message;
}

export class ParsingError extends Error {
	constructor(str: string, parseFailure: ParseFailure) {
		super(generateMessage(str, parseFailure, true));
	}
}
