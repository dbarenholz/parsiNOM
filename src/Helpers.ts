import { ParseResult, STypeBase } from './HelperTypes';
import { Parser } from './Parser';

export function arrayUnion(a: string[] | never[], b: string[] | never[]): string[] {
	const ret: string[] = [...a];
	for (const bElement of b) {
		let alreadyIncluded = false;
		for (const retElement of ret) {
			if (bElement === retElement) {
				alreadyIncluded = true;
			}
		}
		if (!alreadyIncluded) {
			ret.push(bElement);
		}
	}
	ret.sort();
	return ret;
}

// Parser Helpers

export class ParserHelpers {
	followedBy<SType extends STypeBase>(x: Parser<SType>): Parser<SType> {
		return new Parser<SType>(context => {
			const result = x.p(context.copy());
			result.position = context.position;
			return result;
		});
	}

	/**
	 * Returns a parser that yields undefined if the input parser fails, and fails if the input parser accepts. Consumes no input.
	 * More or less an inverse lookahead.
	 *
	 * @param parser
	 */
	notFollowedBy(parser: Parser<unknown>): Parser<undefined> {
		return new Parser((context): ParseResult<undefined> => {
			const contextCopy = context.copy();
			const result = parser.p(contextCopy);
			const text = context.sliceTo(contextCopy.position.index);
			return result.success ? context.fail(`not '` + text + `'`) : context.succeed(undefined);
		});
	}

	/**
	 * Returns a parser that passes the next character into `fn` and yields the character if `fn` returns true.
	 *
	 * @param fn
	 */
	test(fn: (char: string) => boolean): Parser<string> {
		return new Parser<string>((context): ParseResult<string> => {
			const char = context.input[context.position.index];
			if (!context.atEOF() && fn(char)) {
				return context.succeedOffset(1, char);
			} else {
				return context.fail(`a character matching ${fn}`);
			}
		});
	}
}

export const P_HELPERS = new ParserHelpers();
