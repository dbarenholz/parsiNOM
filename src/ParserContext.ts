import { arrayUnion } from './Helpers';
import { ParseResult, ParsingPosition, STypeBase } from './HelperTypes';

export class ParserContext {
	readonly input: string;
	position: ParsingPosition;

	constructor(input: string, position: ParsingPosition) {
		this.input = input;
		this.position = position;
	}

	moveToPosition(position: ParsingPosition): ParserContext {
		this.position = position;
		return this;
	}

	copy(): ParserContext {
		return new ParserContext(this.input, {
			index: this.position.index,
			column: this.position.column,
			line: this.position.line,
		});
	}

	atEOF(): boolean {
		return this.position.index >= this.input.length;
	}

	// OPTIMIZATION: only create a new position when needed, otherwise mutate
	private advanceTo(index: number): ParsingPosition {
		if (index < this.position.index) {
			throw new Error(`Can not advance backwards. Current pos ${this.position.index}. Advance target index ${index}.`);
		}
		if (index === this.position.index) {
			return this.position;
		}

		let line = this.position.line;
		let column = this.position.column;

		for (let i = this.position.index; i < index; i++) {
			if (this.input[i] === '\n') {
				line += 1;
				column = 1;
			} else {
				column += 1;
			}
		}

		this.position = {
			index: index,
			line: line,
			column: column,
		};

		return this.position;
	}

	private invalidPosition(): ParsingPosition {
		return {
			index: -1,
			line: -1,
			column: -1,
		};
	}

	sliceTo(endIndex: number): string {
		return this.input.slice(this.position.index, endIndex);
	}

	succeedOffset<SType extends STypeBase>(offset: number, value: SType): ParseResult<SType> {
		return this.succeedAt(this.position.index + offset, value);
	}

	failOffset<SType extends STypeBase>(offset: number, expected: string | string[]): ParseResult<SType> {
		return this.failAt(this.position.index + offset, expected);
	}

	succeed<SType extends STypeBase>(value: SType): ParseResult<SType> {
		return this.succeedAt(this.position.index, value);
	}

	fail<SType extends STypeBase>(expected: string | string[]): ParseResult<SType> {
		return this.failAt(this.position.index, expected);
	}

	succeedAt<SType extends STypeBase>(index: number, value: SType): ParseResult<SType> {
		return {
			success: true,
			position: this.advanceTo(index),
			value: value,
			furthest: this.invalidPosition(),
			expected: [],
		};
	}

	failAt<SType extends STypeBase>(index: number, expected: string | string[]): ParseResult<SType> {
		return {
			success: false,
			position: this.invalidPosition(),
			value: null,
			furthest: this.advanceTo(index),
			expected: Array.isArray(expected) ? expected : [expected],
		};
	}

	merge<ASType extends STypeBase, BSType extends STypeBase>(a: ParseResult<ASType> | undefined, b: ParseResult<BSType>): ParseResult<BSType> {
		if (a === undefined) {
			return b;
		}

		if (b.furthest.index > a.furthest.index) {
			return b;
		}

		const expected: string[] = b.furthest.index === a.furthest.index ? arrayUnion(a.expected, b.expected) : a.expected;

		b.furthest = a.furthest;
		b.expected = expected;
		return b;
	}
}
