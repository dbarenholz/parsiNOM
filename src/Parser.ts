import { ParserContext } from './ParserContext';
import { ParseFailure, ParseFunction, ParseResult, ParsingMarker, ParsingNode, STypeBase } from './HelperTypes';
import { P, P_UTILS } from './Helpers';

export class Parser<const SType extends STypeBase> {
	public p: ParseFunction<SType>;

	constructor(p: ParseFunction<SType>) {
		this.p = p;
	}

	parse(str: string): ParseResult<SType> {
		return this.p(new ParserContext(str, { index: 0, line: 1, column: 1 }));
	}

	or<OtherSType extends STypeBase>(other: Parser<OtherSType>): Parser<SType | OtherSType> {
		return P.or(this as Parser<SType>, other);
	}

	trim(parser: Parser<unknown>): Parser<SType> {
		return this.wrap(parser, parser);
	}

	wrap(leftParser: Parser<unknown>, rightParser: Parser<unknown>): Parser<SType> {
		return P.sequenceMap(
			(_l, m: SType, _r) => {
				return m;
			},
			leftParser,
			this as Parser<SType>,
			rightParser,
		);
	}

	then<OtherSType extends STypeBase>(next: Parser<OtherSType>): Parser<OtherSType> {
		return P.sequence(this as Parser<SType>, next).map<OtherSType>(results => results[1] as OtherSType);
	}

	many(): Parser<SType[]> {
		return new Parser<SType[]>((context): ParseResult<SType[]> => {
			let result = undefined;
			const startIndex = context.position.index;
			const value: SType[] = [];

			while (true) {
				const iterStartPosition = context.position;
				result = context.merge(result, this.p(context));
				if (result.success) {
					if (result.position.index === startIndex) {
						throw new Error('infinite loop in many() parser detected');
					}
					context.moveToPosition(result.position);
					value.push(result.value);
				} else {
					return context.succeedAtPosition(iterStartPosition, value);
				}
			}
		});
	}

	// TODO: fix this like many. Maybe no fixing is necessary, further investigation needed
	repeat(min: number, max: number): Parser<SType[]> {
		if (max < min) {
			throw new Error('error in times(min, max) parser. max may not be less than min');
		}

		return new Parser<SType[]>((context): ParseResult<SType[]> => {
			let result = undefined;
			let prevResult = undefined;
			const value: SType[] = [];
			let iteration = 0;

			for (; iteration < min; iteration++) {
				result = this.p(context);
				prevResult = context.merge(prevResult, result);
				if (result.success) {
					context.moveToPosition(result.position);
					value.push(result.value);
				} else {
					return prevResult as ParseFailure;
				}
			}

			for (; iteration < max; iteration++) {
				result = this.p(context);
				prevResult = context.merge(prevResult, result);
				if (result.success) {
					context.moveToPosition(result.position);
					value.push(result.value);
				} else {
					break;
				}
			}

			// first move the result to the context position and then change its value.
			return context.succeed(value);
		});
	}

	result<OtherSType extends STypeBase>(value: OtherSType): Parser<OtherSType> {
		return this.map(() => value);
	}

	atMost(max: number): Parser<SType[]> {
		return this.repeat(0, max);
	}

	atLeast(min: number): Parser<SType[]> {
		return P.sequenceMap(
			(part1, part2) => {
				return part1.concat(part2);
			},
			this.repeat(min, min),
			this.many(),
		);
	}

	map<const OtherSType extends STypeBase>(fn: (value: SType) => OtherSType): Parser<OtherSType> {
		return new Parser(context => {
			const result = this.p(context);
			if (!result.success) {
				return result;
			}
			// we are kind of changing the generic type of `ParseResult` here and TS does not like it.
			// return context.mutateResult(result, fn(result.value));

			return context.succeed(fn(result.value));
		});
	}

	skip<const OtherSType extends STypeBase>(next: Parser<OtherSType>): Parser<SType> {
		return P.sequence(this as Parser<SType>, next).map(value => value[0]);
	}

	mark(): Parser<ParsingMarker<SType>> {
		return P.sequenceMap(
			(start, value: SType, end) => {
				return {
					start: start,
					value: value,
					end: end,
				};
			},
			P.pos,
			this as Parser<SType>,
			P.pos,
		);
	}

	node(name: string): Parser<ParsingNode<SType>> {
		return P.sequenceMap(
			(start, value: SType, end) => {
				return {
					name: name,
					start: start,
					value: value,
					end: end,
				};
			},
			P.pos,
			this as Parser<SType>,
			P.pos,
		);
	}

	separateBy(separator: Parser<unknown>): Parser<SType[]> {
		return P.separateBy(this, separator);
	}

	separateByNotEmpty(separator: Parser<unknown>): Parser<SType[]> {
		return P.separateByNotEmpty(this, separator);
	}

	lookahead(x: Parser<SType>): Parser<SType>;
	lookahead(x: string): Parser<string>;
	lookahead(x: RegExp): Parser<string>;
	lookahead(x: Parser<SType> | string | RegExp): Parser<SType | string> {
		return this.skip(P_UTILS.lookahead(x));
	}

	notFollowedBy(next: Parser<unknown>): Parser<SType> {
		return this.skip(P.notFollowedBy(next));
	}

	describe(expected: string | string[]): Parser<SType> {
		if (!Array.isArray(expected)) {
			expected = [expected];
		}

		return new Parser<SType>(context => {
			const result = this.p(context);
			if (!result.success) {
				result.expected = expected as string[];
			}
			return result;
		});
	}

	fallback<OtherSType extends STypeBase>(value: OtherSType): Parser<SType | OtherSType> {
		return this.or(P.alwaysSucceedParser(value));
	}

	chain<OtherSType extends STypeBase>(fn: (result: SType) => Parser<OtherSType>): Parser<OtherSType> {
		return new Parser<OtherSType>(context => {
			const result: ParseResult<SType> = this.p(context);
			if (!result.success) {
				return result;
			}
			const nextParser: Parser<OtherSType> = fn(result.value);
			const nextResult = nextParser.p(context.moveToPosition(result.position));
			return context.merge(result, nextResult);
		});
	}

	optional(): Parser<SType | null> {
		return this.or(P.alwaysSucceedParser(null));
	}
}
