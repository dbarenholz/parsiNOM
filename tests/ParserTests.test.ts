import { P } from '../src/Helpers';
import { testParse } from './TestHelpers';

describe('or', () => {
	const parser = P.string('this').or(P.string('that')).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['that', true],
		['thisthat', false],
		['thatthis', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('trim', () => {
	describe('fixed length', () => {
		const parser = P.string('this').trim(P.string(' ')).skip(P.eof);
		const matchingTable: [string, boolean][] = [
			['', false],
			[' this ', true],
			['this ', false],
			[' this', false],
			['  this', false],
			['foo', false],
		];

		for (const [str, result] of matchingTable) {
			testParse(parser, str, result);
		}
	});

	describe('variable length', () => {
		const parser = P.string('this').trim(P.optWhitespace).skip(P.eof);
		const matchingTable: [string, boolean][] = [
			['', false],
			[' this ', true],
			['this ', true],
			[' this', true],
			['  this', true],
			['foo', false],
		];

		for (const [str, result] of matchingTable) {
			testParse(parser, str, result);
		}
	});
});

describe('wrap', () => {
	const parser = P.string('this').wrap(P.string('('), P.string(')')).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['(this)', true],
		['this)', false],
		['(this', false],
		[')this(', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('then', () => {
	const parser = P.string('this').then(P.string('that')).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['thisthat', true],
		['this', false],
		['that', false],
		['thatthis', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('many', () => {
	const parser = P.string('a').many().skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', true],
		['a', true],
		['aa', true],
		['aaa', true],
		['abaa', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('repeat', () => {
	const parser = P.string('a').repeat(1, 2).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['a', true],
		['aa', true],
		['aaa', false],
		['abaa', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('result', () => {
	const parser = P.string('this').result('result').skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('atMost', () => {
	const parser = P.string('a').atMost(2).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', true],
		['a', true],
		['aa', true],
		['aaa', false],
		['abaa', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('atLeast', () => {
	const parser = P.string('a').atLeast(2).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['a', false],
		['aa', true],
		['aaa', true],
		['abaa', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('map', () => {
	const parser = P.string('this')
		.map(x => x + ' is the result')
		.skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('skip', () => {
	const parser = P.string('this').skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('mark', () => {
	const parser = P.string('this').mark().skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('node', () => {
	const parser = P.string('this').node('fancy node').skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('separateBy', () => {
	const parser = P.string('this').separateBy(P.string(',')).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', true],
		['this', true],
		['this,this', true],
		['this,this,this', true],
		['this, this', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('separateByNotEmpty', () => {
	const parser = P.string('this').separateByNotEmpty(P.string(',')).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['this,this', true],
		['this,this,this', true],
		['this, this', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('lookahead', () => {
	const parser = P.string('this').node('before').lookahead('that').node('after');
	const matchingTable: [string, boolean][] = [
		['', false],
		['thisthat', true],
		['this', false],
		['that', false],
		['thatthis', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('notFollowedBy', () => {
	const parser = P.string('this').node('before').notFollowedBy(P.string('that')).node('after');
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['thisfoo', true],
		['thisthat', false],
		['that', false],
		['thatthis', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('describe', () => {
	const parser = P.string('this').describe('some error').skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('fallback', () => {
	const parser = P.string('this').then(P.string('that').fallback('some fallback')).skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['this', true],
		['thisthat', true],
		['that', false],
		['thatthis', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});

describe('chain', () => {
	const parser = P.string('this')
		.or(P.digit.map(x => Number(x)))
		.chain(res => {
			if (typeof res === 'string') {
				return P.string('that');
			} else {
				return P.string('bar');
			}
		})
		.skip(P.eof);
	const matchingTable: [string, boolean][] = [
		['', false],
		['thisthat', true],
		['1bar', true],
		['thisbar', false],
		['1that', false],
		['this', false],
		['1', false],
		['thatthis', false],
		['foo', false],
	];

	for (const [str, result] of matchingTable) {
		testParse(parser, str, result);
	}
});
