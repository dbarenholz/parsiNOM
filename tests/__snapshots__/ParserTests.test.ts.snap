// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atLeast '' 1`] = `
{
  "expected": [
    "'a'",
  ],
  "value": null,
}
`;

exports[`atLeast 'a' 1`] = `
{
  "expected": [
    "'a'",
  ],
  "value": null,
}
`;

exports[`atLeast 'aa' 1`] = `
{
  "expected": [],
  "value": [
    "a",
    "a",
  ],
}
`;

exports[`atLeast 'aaa' 1`] = `
{
  "expected": [],
  "value": [
    "a",
    "a",
    "a",
  ],
}
`;

exports[`atLeast 'abaa' 1`] = `
{
  "expected": [
    "'a'",
  ],
  "value": null,
}
`;

exports[`atMost '' 1`] = `
{
  "expected": [],
  "value": [],
}
`;

exports[`atMost 'a' 1`] = `
{
  "expected": [],
  "value": [
    "a",
  ],
}
`;

exports[`atMost 'aa' 1`] = `
{
  "expected": [],
  "value": [
    "a",
    "a",
  ],
}
`;

exports[`atMost 'aaa' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`atMost 'abaa' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`chain '' 1`] = `
{
  "expected": [],
  "value": null,
}
`;

exports[`chain '1' 1`] = `
{
  "expected": [
    "'bar'",
  ],
  "value": null,
}
`;

exports[`chain '1bar' 1`] = `
{
  "expected": [],
  "value": "bar",
}
`;

exports[`chain '1that' 1`] = `
{
  "expected": [
    "'bar'",
  ],
  "value": null,
}
`;

exports[`chain 'foo' 1`] = `
{
  "expected": [],
  "value": null,
}
`;

exports[`chain 'thatthis' 1`] = `
{
  "expected": [],
  "value": null,
}
`;

exports[`chain 'this' 1`] = `
{
  "expected": [
    "'that'",
  ],
  "value": null,
}
`;

exports[`chain 'thisbar' 1`] = `
{
  "expected": [
    "'that'",
  ],
  "value": null,
}
`;

exports[`chain 'thisthat' 1`] = `
{
  "expected": [],
  "value": "that",
}
`;

exports[`describe '' 1`] = `
{
  "expected": [
    "some error",
  ],
  "value": null,
}
`;

exports[`describe 'foo' 1`] = `
{
  "expected": [
    "some error",
  ],
  "value": null,
}
`;

exports[`describe 'this' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`fallback '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`fallback 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`fallback 'that' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`fallback 'thatthis' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`fallback 'this' 1`] = `
{
  "expected": [],
  "value": "some fallback",
}
`;

exports[`fallback 'thisthat' 1`] = `
{
  "expected": [],
  "value": "that",
}
`;

exports[`lookahead '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`lookahead 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`lookahead 'that' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`lookahead 'thatthis' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`lookahead 'this' 1`] = `
{
  "expected": [
    "'that'",
  ],
  "value": "",
}
`;

exports[`lookahead 'thisthat' 1`] = `
{
  "expected": [],
  "value": {
    "end": {
      "column": 5,
      "index": 4,
      "line": 1,
    },
    "name": "after",
    "start": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "value": {
      "end": {
        "column": 5,
        "index": 4,
        "line": 1,
      },
      "name": "before",
      "start": {
        "column": 1,
        "index": 0,
        "line": 1,
      },
      "value": "this",
    },
  },
}
`;

exports[`many '' 1`] = `
{
  "expected": [],
  "value": [],
}
`;

exports[`many 'a' 1`] = `
{
  "expected": [],
  "value": [
    "a",
  ],
}
`;

exports[`many 'aa' 1`] = `
{
  "expected": [],
  "value": [
    "a",
    "a",
  ],
}
`;

exports[`many 'aaa' 1`] = `
{
  "expected": [],
  "value": [
    "a",
    "a",
    "a",
  ],
}
`;

exports[`many 'abaa' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`map '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`map 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`map 'this' 1`] = `
{
  "expected": [],
  "value": "this is the result",
}
`;

exports[`mark '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`mark 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`mark 'this' 1`] = `
{
  "expected": [],
  "value": {
    "end": {
      "column": 5,
      "index": 4,
      "line": 1,
    },
    "start": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "value": "this",
  },
}
`;

exports[`node '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`node 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`node 'this' 1`] = `
{
  "expected": [],
  "value": {
    "end": {
      "column": 5,
      "index": 4,
      "line": 1,
    },
    "name": "fancy node",
    "start": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "value": "this",
  },
}
`;

exports[`notFollowedBy '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`notFollowedBy 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`notFollowedBy 'that' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`notFollowedBy 'thatthis' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`notFollowedBy 'this' 1`] = `
{
  "expected": [],
  "value": {
    "end": {
      "column": 5,
      "index": 4,
      "line": 1,
    },
    "name": "after",
    "start": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "value": {
      "end": {
        "column": 5,
        "index": 4,
        "line": 1,
      },
      "name": "before",
      "start": {
        "column": 1,
        "index": 0,
        "line": 1,
      },
      "value": "this",
    },
  },
}
`;

exports[`notFollowedBy 'thisfoo' 1`] = `
{
  "expected": [],
  "value": {
    "end": {
      "column": 5,
      "index": 4,
      "line": 1,
    },
    "name": "after",
    "start": {
      "column": 1,
      "index": 0,
      "line": 1,
    },
    "value": {
      "end": {
        "column": 5,
        "index": 4,
        "line": 1,
      },
      "name": "before",
      "start": {
        "column": 1,
        "index": 0,
        "line": 1,
      },
      "value": "this",
    },
  },
}
`;

exports[`notFollowedBy 'thisthat' 1`] = `
{
  "expected": [
    "not 'that'",
  ],
  "value": null,
}
`;

exports[`or '' 1`] = `
{
  "expected": [],
  "value": null,
}
`;

exports[`or 'foo' 1`] = `
{
  "expected": [],
  "value": null,
}
`;

exports[`or 'that' 1`] = `
{
  "expected": [],
  "value": "that",
}
`;

exports[`or 'thatthis' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`or 'this' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`or 'thisthat' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`repeat '' 1`] = `
{
  "expected": [
    "'a'",
  ],
  "value": null,
}
`;

exports[`repeat 'a' 1`] = `
{
  "expected": [],
  "value": [
    "a",
  ],
}
`;

exports[`repeat 'aa' 1`] = `
{
  "expected": [],
  "value": [
    "a",
    "a",
  ],
}
`;

exports[`repeat 'aaa' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`repeat 'abaa' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`result '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`result 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`result 'this' 1`] = `
{
  "expected": [],
  "value": "result",
}
`;

exports[`separateBy '' 1`] = `
{
  "expected": [],
  "value": [],
}
`;

exports[`separateBy 'foo' 1`] = `
{
  "expected": [],
  "value": null,
}
`;

exports[`separateBy 'this' 1`] = `
{
  "expected": [],
  "value": [
    "this",
  ],
}
`;

exports[`separateBy 'this, this' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`separateBy 'this,this' 1`] = `
{
  "expected": [],
  "value": [
    "this",
    "this",
  ],
}
`;

exports[`separateBy 'this,this,this' 1`] = `
{
  "expected": [],
  "value": [
    "this",
    "this",
    "this",
  ],
}
`;

exports[`separateByNotEmpty '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`separateByNotEmpty 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`separateByNotEmpty 'this' 1`] = `
{
  "expected": [],
  "value": [
    "this",
  ],
}
`;

exports[`separateByNotEmpty 'this, this' 1`] = `
{
  "expected": [
    "eof",
  ],
  "value": null,
}
`;

exports[`separateByNotEmpty 'this,this' 1`] = `
{
  "expected": [],
  "value": [
    "this",
    "this",
  ],
}
`;

exports[`separateByNotEmpty 'this,this,this' 1`] = `
{
  "expected": [],
  "value": [
    "this",
    "this",
    "this",
  ],
}
`;

exports[`skip '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`skip 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`skip 'this' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`then '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`then 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`then 'that' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`then 'thatthis' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`then 'this' 1`] = `
{
  "expected": [
    "'that'",
  ],
  "value": null,
}
`;

exports[`then 'thisthat' 1`] = `
{
  "expected": [],
  "value": "that",
}
`;

exports[`trim fixed length '  this' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`trim fixed length ' this ' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`trim fixed length ' this' 1`] = `
{
  "expected": [
    "' '",
  ],
  "value": null,
}
`;

exports[`trim fixed length '' 1`] = `
{
  "expected": [
    "' '",
  ],
  "value": null,
}
`;

exports[`trim fixed length 'foo' 1`] = `
{
  "expected": [
    "' '",
  ],
  "value": null,
}
`;

exports[`trim fixed length 'this ' 1`] = `
{
  "expected": [
    "' '",
  ],
  "value": null,
}
`;

exports[`trim variable length '  this' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`trim variable length ' this ' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`trim variable length ' this' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`trim variable length '' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`trim variable length 'foo' 1`] = `
{
  "expected": [
    "'this'",
  ],
  "value": null,
}
`;

exports[`trim variable length 'this ' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`wrap '' 1`] = `
{
  "expected": [
    "'('",
  ],
  "value": null,
}
`;

exports[`wrap '(this' 1`] = `
{
  "expected": [
    "')'",
  ],
  "value": null,
}
`;

exports[`wrap '(this)' 1`] = `
{
  "expected": [],
  "value": "this",
}
`;

exports[`wrap ')this(' 1`] = `
{
  "expected": [
    "'('",
  ],
  "value": null,
}
`;

exports[`wrap 'foo' 1`] = `
{
  "expected": [
    "'('",
  ],
  "value": null,
}
`;

exports[`wrap 'this)' 1`] = `
{
  "expected": [
    "'('",
  ],
  "value": null,
}
`;
